### PR TITLE
Add Linux GameMode support

### PR DESCRIPTION
**Pull request by realnc**
## Description

This can fix a lot of performance issues, like audio crackling and frame
time spikes. This requires the GameMode package to be installed. See:

https://github.com/FeralInteractive/gamemode

This commit adds a "Game Mode" bool option to the "Power
Management" and "Latency" settings sections, and it can be toggled
on/off without restarting RA.

The actual toggling of game mode happens in a new frontend platform
interface function. Perhaps this will become useful for other platforms
that provide some equivalent of Linux GameMode.

Since the GameMode ABI is fixed, and the API comes as a single,
header-only file with no actual deps, we simply bundle the header
(deps/feralgamemode/gamemode_client.h.) That way, all Linux builds will
have support for GameMode regardless of whether the GameMode development
package is installed or not.

## Reviewers

@jdgleaver @twinaphex

### CODE REVIEW

**Comment by realnc** [comment]
I'm not sure how to handle the `gamemode_client.h` dependency. Right now, it's just found with pkg-config and so it will need to be installed on the build bot. However, that header can also be bundled and used unconditionally. In that case, GameMode will simply not work on systems that don't have that software installed (the `gamemode_client.h` functions use `dlopen`.)

Thoughts?

File: menu/menu_setting.c [hunk 1]
```diff
@@ -7594,6 +7598,22 @@ static void general_write_handler(rarch_setting_t *setting)
                task_queue_unset_threaded();
          }
          break;
+#ifdef HAVE_INUXGAMEMODE
```
Comment (RobLoach) [inline menu/menu_setting.c:7601]:
``` suggestion
#ifdef HAVE_LINUXGAMEMODE
```

File: menu/menu_setting.c [hunk 2]
```diff
@@ -7594,6 +7598,22 @@ static void general_write_handler(rarch_setting_t *setting)
...
+      case MENU_ENUM_LABEL_LINUX_GAMEMODE_ENABLE:
+         if (config_get_ptr()->bools.linux_gamemode_enable && gamemode_query_status() != 2)
+         {
+            if (gamemode_request_start() != 0) {
+               RARCH_WARN("Failed to enter game mode: %s.\n", gamemode_error_string());
```
Comment (RobLoach) [inline menu/menu_setting.c:7606]:
``` suggestion
               RARCH_WARN("[GameMode]: Failed to enter game mode: %s.\n", gamemode_error_string());
```

File: menu/menu_setting.c [hunk 3]
```diff
@@ -7594,6 +7598,22 @@ static void general_write_handler(rarch_setting_t *setting)
...
+            }
+         }
+         else if (!config_get_ptr()->bools.linux_gamemode_enable && gamemode_query_status() == 2)
+         {
+            if (gamemode_request_end() != 0) {
+               RARCH_WARN("Failed to exit game mode: %s.\n", gamemode_error_string());
```
Comment (RobLoach) [inline menu/menu_setting.c:7612]:
``` suggestion
               RARCH_WARN("[GameMode]: Failed to exit game mode: %s.\n", gamemode_error_string());
```

File: retroarch.c [hunk 4]
```diff
@@ -3813,6 +3823,16 @@ int rarch_main(int argc, char *argv[], void *data)
    }
 
    ui_companion_driver_init_first();
+
+#ifdef HAVE_LINUXGAMEMODE
+   if (
+      config_get_ptr()->bools.linux_gamemode_enable && gamemode_query_status() != 2
+      && gamemode_request_start() != 0)
+   {
+      RARCH_WARN("Failed to enter game mode: %s.\n", gamemode_error_string());
```
Comment (RobLoach) [inline retroarch.c:3832]:
``` suggestion
      RARCH_WARN("[GameMode]: Failed to enter game mode: %s.\n", gamemode_error_string());
```

File: retroarch.c [hunk 5]
```diff
@@ -3673,6 +3677,12 @@ void main_exit(void *args)
       runloop_log_counters(p_rarch->perf_counters_rarch, p_rarch->perf_ptr_rarch);
    }
 
+#ifdef HAVE_LINUXGAMEMODE
+   if (gamemode_query_status() == 2 && gamemode_request_end() != 0) {
+      RARCH_WARN("Failed to exit game mode: %s.\n", gamemode_error_string());
```
Comment (RobLoach) [inline retroarch.c:3682]:
``` suggestion
      RARCH_WARN("[GameMode]: Failed to exit game mode: %s.\n", gamemode_error_string());
```

File: config.def.h [hunk 6]
```diff
@@ -173,6 +173,8 @@
 
 #define DEFAULT_USER_LANGUAGE 0
 
+#define DEFAULT_LINUX_GAMEMODE_ENABLE true
```
Comment (RobLoach) [inline config.def.h:176]: Not sure if we should enable GameMode by default. While it shouldn't have an impact on user experience, I think having the user opt-in may be the right move. Don't want to do something that they don't expect.

``` suggestion
#define DEFAULT_LINUX_GAMEMODE_ENABLE false
```

**Comment by RobLoach** [comment]
> I'm not sure how to handle the gamemode_client.h dependency. Right now, it's just found with pkg-config and so it will need to be installed on the build bot. However, that header can also be bundled and used unconditionally. In that case, GameMode will simply not work on systems that don't have that software installed (the gamemode_client.h functions use dlopen.)

I think how it is seems reasonable. I didn't have `libgamemode-dev` installed, and it was building without gamemode enabled. Once libgamemode-dev was installed and re-compiled, it worked. When I removed gamemode from my system, I got a reasonable error:
```
[WARN] GameMode: Failed to enter game mode: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory.
```

File: config.def.h (hunk 6 above)
Comment (realnc) [reply config.def.h:176]: I think it should be on by default. This is how Windows handles its own game mode (on unless a game is blacklisted or not recognized as a game.) It's also what games that support GameMode do. If available, they enable it. And I don't think they even offer an option to disable.

I'd say that if a user installs GameMode, they want it used plug&play style without having to go over every single game they have and explicitly enable it, which would be kind of annoying :-)

**Comment by realnc** [comment]
> I think how it is seems reasonable. I didn't have `libgamemode-dev` installed, and it was building without gamemode enabled. Once libgamemode-dev was installed and re-compiled, it worked. When I removed gamemode from my system, I got a reasonable error:
> 
> ```
> [WARN] GameMode: Failed to enter game mode: dlopen failed - libgamemode.so: cannot open shared object file: No such file or directory.
> ```

The problem is that it's not installed on the build bot. So when people download RA instead of building it themselves, they get a build that doesn't support gamemode.

So there's two choices. Make sure it's installed on the build bot, or bundle the header.

**Comment by RobLoach** [comment]
True... Perhaps let the builds and package managers determine whether GameMode is available? If the header isn't there, assume the build doesn't have it seems fine. I could likely get it up and running in Flatpak and Snap.

**Comment by jdgleaver** [comment]
@realnc Thank you, this is excellent :)

For proper correctness, however, the gamemode interface should be accessed inside the unix frontend platform driver - and it seems that the `frontend_unix_set_sustained_performance_mode()` function would be the ideal place for this. We can in fact use all the existing 'sustained performance' code, and just sub in different menu entry text depending upon whether we're running on Android or Linux. I will prepare a little patch to implement this change, and let you review it :)

As for your other points:

- Installing `libgamemode-dev` on the buildbot seems preferable to bundling the header. I can address this once the PR is merged

- I agree that gamemode should ideally be enabled by default for self-builds - but this makes things sticky for users who download a prebuilt version. Most people won't have gamemode installed, so they're all going to get dlopen failures and log spam. We perhaps need a command line flag to set it disabled by default for the buildbot versions...

**Comment by RobLoach** [comment]
> Most people won't have gamemode installed, so they're all going to get dlopen failures and log spam. 

This is why I believe having it disabled by default may be preferable.

**Comment by realnc** [comment]
The dlopen failure messages are fine. After all, I constantly get wayland errors with no way to disable them. So nobody really cares about these messages, I'd say. And in the case of gamemode, unlike wayland, you can at least disable it.

**Comment by realnc** [comment]
> the `frontend_unix_set_sustained_performance_mode()` function would be the ideal place for this.

Nah, gamemode is also about changing the CPU scheduler, I/O priority, nice level, and running custom scripts the user has provided.

Gamemode is it's own thing and should be treated as its own thing in the source code, IMO. I would never expect to find gamemode related code in that function.

**Comment by jdgleaver** [comment]
> Nah, gamemode is also about changing the CPU scheduler, I/O priority, nice level, and running custom scripts the user has provided.
>
> Gamemode is it's own thing and should be treated as its own thing in the source code, IMO. I would never expect to find gamemode related code in that function.

You missed my point. This *has* to go inside the unix frontend platform driver. This means it has to be enabled/disabled via a frontend platform interface function. We already have a 'sustained performance' function available on all platforms. If you don't want to put it in there, then you'll have to add a new frontend platform interface function.

**Comment by realnc** [comment]
> You missed my point. This _has_ to go inside the unix frontend platform driver. This means it has to be enabled/disabled via a frontend platform interface function. We already have a 'sustained performance' function available on all platforms. If you don't want to put it in there, then you'll have to add a new frontend platform interface function.

Thanks. That makes sense. I'll add a new interface function for this.

**Comment by realnc** [comment]
@jdgleaver I reworked the PR and it now uses a new platform interface function.

**Comment by jdgleaver** [comment]
Okay, thanks. A couple of things remain:

- You need to call `frontend_unix_set_linux_gamemode(false)` in `frontend_unix_deinit()` if `HAVE_LINUXGAMEMODE` is set

- Calling the frontend interface function `set_linux_gamemode` is somewhat nonsensical, isn't it? Why would we have a Linux-specific function in the frontend drivers for every platform? This should instead be a generic 'game mode' function, which could potentially be used on other platforms (and as such is very much like the existing 'sustained performance' interface function, as I said before...). In this case, all 'linux' references should be removed from the code, with only the menu entry sublabel (and info text) changing to reflect the host platform. In addition, there is no longer a need to gate any of the code outside the frontend driver file behind `HAVE_LINUXGAMEMODE` ifdefs - you can instead show menu options dynamically based on whether the frontend 'game mode' function is non-NULL

**Comment by realnc** [comment]
I did the changes. One thing though:

> all 'linux' references should be removed from the code, with only the menu entry sublabel (and info text) changing to reflect the host platform.

I don't know how to do that for the sublabel text.

**Comment by RobLoach** [comment]
> I don't know how to do that for the sublabel text.

Perhaps remove all mention of "Linux" throughout then? It's branded as "GameMode", so I think that should be enough.

**Comment by realnc** [comment]
> > I don't know how to do that for the sublabel text.
> 
> Perhaps remove all mention of "Linux" throughout then? It's branded as "GameMode", so I think that should be enough.

It's fixed now. Thanks everyone for all the help.

Also, since the GameMode ABI is fixed (commercial games out there already use it and it's never gonna break,) the header is now bundled. It's header-only with no deps.

So this PR is now ready to merge.

**Comment by realnc** [comment]
Oh, the C89 test failed. The header uses `//` comments :P

**Comment by realnc** [comment]
OK, should be all good now for merging. Game mode support is now disabled when building in C89 mode. A system so old that needs C89 mode can't have game mode to begin with, since the game mode software itself requires at least C99.

**Review by jdgleaver — APPROVED** [review]

**Comment by jdgleaver** [comment]
Excellent - this all looks good now.

**Pull request merged**